
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>&lt;no title&gt; &#8212; Open Science Data Repository (OSDR) NoteBook Tutorial</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.1/css/all.min.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.1/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.1/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.1/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=384b581d" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css?v=0a3b3ea7" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae" />
  <script src="_static/vendor/fontawesome/6.5.1/js/all.min.js?digest=8d27b9dea8ad943066ae"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=efea14e4"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=36754332"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'Knowledge-graph-notebook_for_ISSOP_figures';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Table of contents" href="SUMMARY.html" />
    <link rel="prev" title="&lt;no title&gt;" href="PubMed_scraping_notenook_for_ISSOP_figures.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a id="pst-skip-link" class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <header class="bd-header navbar navbar-expand-lg bd-navbar">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  

<a class="navbar-brand logo" href="intro.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/GeneLab_logo2.png" class="logo__image only-light" alt="Open Science Data Repository (OSDR) NoteBook Tutorial - Home"/>
    <script>document.write(`<img src="_static/GeneLab_logo2.png" class="logo__image only-dark" alt="Open Science Data Repository (OSDR) NoteBook Tutorial - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Welcome to your Jupyter Book
                </a>
            </li>
        </ul>
        <ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="Methods_Lesson_01.html">AI/ML methods notebook</a></li>







<li class="toctree-l1"><a class="reference internal" href="rr9-phenotypes_Lesson_01.html">RR9 phenotypes notebook summary</a></li>




<li class="toctree-l1"><a class="reference internal" href="Methods.html">AI/ML methods notebook</a></li>









<li class="toctree-l1"><a class="reference internal" href="Regression.html"><strong>Mission of regression notebook</strong></a></li>






<li class="toctree-l1"><a class="reference internal" href="Classification.html"><strong>Mission of the classification notebook</strong></a></li>




<li class="toctree-l1"><a class="reference internal" href="Clustering.html"><strong>Mission of clustering notebook</strong></a></li>







<li class="toctree-l1"><a class="reference internal" href="Image_Data.html">Introduction to Image Analysis with Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="SUMMARY.html">Table of contents</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/dr-richard-barker/OSDR_juypter_book.io" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/dr-richard-barker/OSDR_juypter_book.io/issues/new?title=Issue%20on%20page%20%2FKnowledge-graph-notebook_for_ISSOP_figures.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/Knowledge-graph-notebook_for_ISSOP_figures.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1><no title></h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="simple visible nav section-nav flex-column">
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.gridspec</span> <span class="k">as</span> <span class="nn">gridspec</span>
<span class="kn">import</span> <span class="nn">matplotlib.patches</span> <span class="k">as</span> <span class="nn">patches</span>

<span class="c1"># Load the data</span>
<span class="n">file_path</span> <span class="o">=</span> <span class="s1">&#39;genelab_osd.txt&#39;</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="c1"># Minor error/formatting corrections to data</span>
<span class="n">data</span><span class="p">[</span><span class="s1">&#39;technologyType&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;technologyType&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;single-cell RNA Sequencing&#39;</span><span class="p">,</span> <span class="s1">&#39;single-cell RNA sequencing&#39;</span><span class="p">)</span>

<span class="n">data</span><span class="p">[</span><span class="s1">&#39;technologyType&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;technologyType&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;DNA microarray&#39;</span><span class="p">,</span> <span class="s1">&#39;microarray&#39;</span><span class="p">)</span>

<span class="c1"># Helper function to sort measures into omics categories</span>
<span class="k">def</span> <span class="nf">assign_category</span><span class="p">(</span><span class="n">measurement_type</span><span class="p">):</span>
    <span class="n">categories</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;genomics&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;genome sequencing&quot;</span><span class="p">,</span> <span class="s2">&quot;copy number variation profiling&quot;</span><span class="p">,</span>
                         <span class="s2">&quot;deletion pool profiling&quot;</span><span class="p">,</span> <span class="s2">&quot;Amplicon Sequencing&quot;</span><span class="p">,</span>
                         <span class="s2">&quot;Genome-wide Association Study&quot;</span><span class="p">],</span>
            <span class="s1">&#39;transcriptomics&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;transcription profiling&quot;</span><span class="p">,</span> <span class="s2">&quot;RNA methylation profiling&quot;</span><span class="p">],</span>
            <span class="s1">&#39;proteomics&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;protein expression profiling&quot;</span><span class="p">,</span> <span class="s2">&quot;protein quantification&quot;</span><span class="p">,</span>
                           <span class="s2">&quot;phosphoprotein profiling&quot;</span><span class="p">,</span> <span class="s2">&quot;protein-DNA binding site identification&quot;</span><span class="p">,</span>
                           <span class="s2">&quot;protein identification&quot;</span><span class="p">],</span>
            <span class="s1">&#39;metabolomics&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;metabolite profiling&quot;</span><span class="p">],</span>
            <span class="s1">&#39;epigenomics&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;DNA methylation profiling&quot;</span><span class="p">,</span> <span class="s2">&quot;Chromatin Accessibility&quot;</span><span class="p">],</span>
            <span class="s1">&#39;metagenomics&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;metagenome profiling&quot;</span><span class="p">,</span>
                                                          <span class="s2">&quot;Metagenomic sequencing&quot;</span><span class="p">,</span>
                                                          <span class="s2">&quot;environmental gene survey&quot;</span><span class="p">],</span>
            <span class="s1">&#39;non-omics&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Molecular Cellular Imaging&quot;</span><span class="p">,</span> <span class="s2">&quot;Behavior&quot;</span><span class="p">,</span> <span class="s2">&quot;Bone Microstructure&quot;</span><span class="p">,</span>
                          <span class="s2">&quot;Echocardiogram&quot;</span><span class="p">,</span> <span class="s2">&quot;Histological Assay&quot;</span><span class="p">,</span> <span class="s2">&quot;Image Analysis&quot;</span><span class="p">,</span>
                          <span class="s2">&quot;Morphometric analysis&quot;</span><span class="p">,</span> <span class="s2">&quot;Bone biomechanical measurement&quot;</span><span class="p">,</span>
                          <span class="s2">&quot;Calcium Uptake&quot;</span><span class="p">,</span> <span class="s2">&quot;histology&quot;</span><span class="p">,</span> <span class="s2">&quot;Intraocular Pressure&quot;</span><span class="p">,</span> <span class="s2">&quot;Flow Cytometry&quot;</span><span class="p">]</span>
        <span class="p">}</span>

    <span class="k">for</span> <span class="n">category</span><span class="p">,</span> <span class="n">measurements</span> <span class="ow">in</span> <span class="n">categories</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">measurement_type</span> <span class="ow">in</span> <span class="n">measurements</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">category</span>

    <span class="k">return</span> <span class="s2">&quot;Uncategorized&quot;</span>

<span class="c1"># Assign an omics type to each row based on measurementType</span>
<span class="n">data</span><span class="p">[</span><span class="s1">&#39;Category&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;measurementType&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">assign_category</span><span class="p">)</span>

<span class="c1"># category counts</span>
<span class="nb">print</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;Category&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">())</span>


<span class="c1"># this is same classification that Henry decidid for his figure</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">FileNotFoundError</span><span class="g g-Whitespace">                         </span>Traceback (most recent call last)
<span class="n">Cell</span> <span class="n">In</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">line</span> <span class="mi">9</span>
<span class="g g-Whitespace">      </span><span class="mi">7</span> <span class="c1"># Load the data</span>
<span class="g g-Whitespace">      </span><span class="mi">8</span> <span class="n">file_path</span> <span class="o">=</span> <span class="s1">&#39;genelab_osd.txt&#39;</span>
<span class="ne">----&gt; </span><span class="mi">9</span> <span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">11</span> <span class="c1"># Minor error/formatting corrections to data</span>
<span class="g g-Whitespace">     </span><span class="mi">12</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;technologyType&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;technologyType&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;single-cell RNA Sequencing&#39;</span><span class="p">,</span> <span class="s1">&#39;single-cell RNA sequencing&#39;</span><span class="p">)</span>

<span class="nn">File /workspace/.pyenv_mirror/user/current/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1026,</span> in <span class="ni">read_csv</span><span class="nt">(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)</span>
<span class="g g-Whitespace">   </span><span class="mi">1013</span> <span class="n">kwds_defaults</span> <span class="o">=</span> <span class="n">_refine_defaults_read</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1014</span>     <span class="n">dialect</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1015</span>     <span class="n">delimiter</span><span class="p">,</span>
   <span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1022</span>     <span class="n">dtype_backend</span><span class="o">=</span><span class="n">dtype_backend</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1023</span> <span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1024</span> <span class="n">kwds</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">kwds_defaults</span><span class="p">)</span>
<span class="ne">-&gt; </span><span class="mi">1026</span> <span class="k">return</span> <span class="n">_read</span><span class="p">(</span><span class="n">filepath_or_buffer</span><span class="p">,</span> <span class="n">kwds</span><span class="p">)</span>

<span class="nn">File /workspace/.pyenv_mirror/user/current/lib/python3.12/site-packages/pandas/io/parsers/readers.py:620,</span> in <span class="ni">_read</span><span class="nt">(filepath_or_buffer, kwds)</span>
<span class="g g-Whitespace">    </span><span class="mi">617</span> <span class="n">_validate_names</span><span class="p">(</span><span class="n">kwds</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;names&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">))</span>
<span class="g g-Whitespace">    </span><span class="mi">619</span> <span class="c1"># Create the parser.</span>
<span class="ne">--&gt; </span><span class="mi">620</span> <span class="n">parser</span> <span class="o">=</span> <span class="n">TextFileReader</span><span class="p">(</span><span class="n">filepath_or_buffer</span><span class="p">,</span> <span class="o">**</span><span class="n">kwds</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">622</span> <span class="k">if</span> <span class="n">chunksize</span> <span class="ow">or</span> <span class="n">iterator</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">623</span>     <span class="k">return</span> <span class="n">parser</span>

<span class="nn">File /workspace/.pyenv_mirror/user/current/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1620,</span> in <span class="ni">TextFileReader.__init__</span><span class="nt">(self, f, engine, **kwds)</span>
<span class="g g-Whitespace">   </span><span class="mi">1617</span>     <span class="bp">self</span><span class="o">.</span><span class="n">options</span><span class="p">[</span><span class="s2">&quot;has_index_names&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">kwds</span><span class="p">[</span><span class="s2">&quot;has_index_names&quot;</span><span class="p">]</span>
<span class="g g-Whitespace">   </span><span class="mi">1619</span> <span class="bp">self</span><span class="o">.</span><span class="n">handles</span><span class="p">:</span> <span class="n">IOHandles</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span>
<span class="ne">-&gt; </span><span class="mi">1620</span> <span class="bp">self</span><span class="o">.</span><span class="n">_engine</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_make_engine</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">engine</span><span class="p">)</span>

<span class="nn">File /workspace/.pyenv_mirror/user/current/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1880,</span> in <span class="ni">TextFileReader._make_engine</span><span class="nt">(self, f, engine)</span>
<span class="g g-Whitespace">   </span><span class="mi">1878</span>     <span class="k">if</span> <span class="s2">&quot;b&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">mode</span><span class="p">:</span>
<span class="g g-Whitespace">   </span><span class="mi">1879</span>         <span class="n">mode</span> <span class="o">+=</span> <span class="s2">&quot;b&quot;</span>
<span class="ne">-&gt; </span><span class="mi">1880</span> <span class="bp">self</span><span class="o">.</span><span class="n">handles</span> <span class="o">=</span> <span class="n">get_handle</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1881</span>     <span class="n">f</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1882</span>     <span class="n">mode</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1883</span>     <span class="n">encoding</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">options</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;encoding&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">),</span>
<span class="g g-Whitespace">   </span><span class="mi">1884</span>     <span class="n">compression</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">options</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;compression&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">),</span>
<span class="g g-Whitespace">   </span><span class="mi">1885</span>     <span class="n">memory_map</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">options</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;memory_map&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">),</span>
<span class="g g-Whitespace">   </span><span class="mi">1886</span>     <span class="n">is_text</span><span class="o">=</span><span class="n">is_text</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1887</span>     <span class="n">errors</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">options</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;encoding_errors&quot;</span><span class="p">,</span> <span class="s2">&quot;strict&quot;</span><span class="p">),</span>
<span class="g g-Whitespace">   </span><span class="mi">1888</span>     <span class="n">storage_options</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">options</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;storage_options&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">),</span>
<span class="g g-Whitespace">   </span><span class="mi">1889</span> <span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1890</span> <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">handles</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
<span class="g g-Whitespace">   </span><span class="mi">1891</span> <span class="n">f</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">handles</span><span class="o">.</span><span class="n">handle</span>

<span class="nn">File /workspace/.pyenv_mirror/user/current/lib/python3.12/site-packages/pandas/io/common.py:873,</span> in <span class="ni">get_handle</span><span class="nt">(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)</span>
<span class="g g-Whitespace">    </span><span class="mi">868</span> <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
<span class="g g-Whitespace">    </span><span class="mi">869</span>     <span class="c1"># Check whether the filename is to be opened in binary mode.</span>
<span class="g g-Whitespace">    </span><span class="mi">870</span>     <span class="c1"># Binary mode does not support &#39;encoding&#39; and &#39;newline&#39;.</span>
<span class="g g-Whitespace">    </span><span class="mi">871</span>     <span class="k">if</span> <span class="n">ioargs</span><span class="o">.</span><span class="n">encoding</span> <span class="ow">and</span> <span class="s2">&quot;b&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ioargs</span><span class="o">.</span><span class="n">mode</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">872</span>         <span class="c1"># Encoding</span>
<span class="ne">--&gt; </span><span class="mi">873</span>         <span class="n">handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span>
<span class="g g-Whitespace">    </span><span class="mi">874</span>             <span class="n">handle</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">875</span>             <span class="n">ioargs</span><span class="o">.</span><span class="n">mode</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">876</span>             <span class="n">encoding</span><span class="o">=</span><span class="n">ioargs</span><span class="o">.</span><span class="n">encoding</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">877</span>             <span class="n">errors</span><span class="o">=</span><span class="n">errors</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">878</span>             <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">879</span>         <span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">880</span>     <span class="k">else</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">881</span>         <span class="c1"># Binary mode</span>
<span class="g g-Whitespace">    </span><span class="mi">882</span>         <span class="n">handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span> <span class="n">ioargs</span><span class="o">.</span><span class="n">mode</span><span class="p">)</span>

<span class="ne">FileNotFoundError</span>: [Errno 2] No such file or directory: &#39;genelab_osd.txt&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>OSD-id</th>
      <th>hits</th>
      <th>publicReleaseDate</th>
      <th>publicReleaseDateYear</th>
      <th>technologyPlatform</th>
      <th>technologyType</th>
      <th>measurementType</th>
      <th>OrganismCount</th>
      <th>Organism</th>
      <th>Factors</th>
      <th>Category</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>OSD-1</td>
      <td>1</td>
      <td>11-Dec-2013</td>
      <td>2013</td>
      <td>Affymetrix</td>
      <td>microarray</td>
      <td>transcription profiling</td>
      <td>1</td>
      <td>Drosophila melanogaster</td>
      <td>Spaceflight;Infection</td>
      <td>transcriptomics</td>
    </tr>
    <tr>
      <th>1</th>
      <td>OSD-2</td>
      <td>1</td>
      <td>01-Apr-2011</td>
      <td>2011</td>
      <td>Agilent</td>
      <td>microarray</td>
      <td>transcription profiling</td>
      <td>1</td>
      <td>Homo sapiens</td>
      <td>Ionizing Radiation;Absorbed Radiation Dose</td>
      <td>transcriptomics</td>
    </tr>
    <tr>
      <th>2</th>
      <td>OSD-3</td>
      <td>1</td>
      <td>01-Sep-2011</td>
      <td>2011</td>
      <td>Affymetrix</td>
      <td>microarray</td>
      <td>transcription profiling</td>
      <td>1</td>
      <td>Drosophila melanogaster</td>
      <td>Developmental Stage;Spaceflight;Cosmic Radiation</td>
      <td>transcriptomics</td>
    </tr>
    <tr>
      <th>3</th>
      <td>OSD-4</td>
      <td>1</td>
      <td>10-Mar-2010</td>
      <td>2010</td>
      <td>Affymetrix</td>
      <td>microarray</td>
      <td>transcription profiling</td>
      <td>1</td>
      <td>Mus musculus</td>
      <td>Spaceflight</td>
      <td>transcriptomics</td>
    </tr>
    <tr>
      <th>4</th>
      <td>OSD-5</td>
      <td>1</td>
      <td>09-Feb-2006</td>
      <td>2006</td>
      <td>Affymetrix</td>
      <td>microarray</td>
      <td>transcription profiling</td>
      <td>1</td>
      <td>Homo sapiens</td>
      <td>simulated microgravity</td>
      <td>transcriptomics</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>540</th>
      <td>OSD-628</td>
      <td>1</td>
      <td>12-Sep-2023</td>
      <td>2023</td>
      <td>Nikon SIM A1 Laser scanning Confocal Microscope</td>
      <td>Confocal</td>
      <td>Molecular Cellular Imaging</td>
      <td>1</td>
      <td>Penicillium rubens</td>
      <td>Spaceflight;Growth Environment;Time</td>
      <td>non-omics</td>
    </tr>
    <tr>
      <th>541</th>
      <td>OSD-640</td>
      <td>1</td>
      <td>24-Aug-2023</td>
      <td>2023</td>
      <td>FACS Canto2(Becton Dickinson)</td>
      <td>Flow Cytometry</td>
      <td>Flow Cytometry</td>
      <td>1</td>
      <td>Mus musculus</td>
      <td>Hindlimb Unloading;Sex;Housing Condition;Ioniz...</td>
      <td>non-omics</td>
    </tr>
    <tr>
      <th>542</th>
      <td>OSD-654</td>
      <td>2</td>
      <td>04-Oct-2023</td>
      <td>2023</td>
      <td>SkyScan 1176 V1 build 12</td>
      <td>Micro-Computed Tomography</td>
      <td>Bone Microstructure</td>
      <td>1</td>
      <td>Mus musculus</td>
      <td>Spaceflight;Space Mission;Genotype</td>
      <td>non-omics</td>
    </tr>
    <tr>
      <th>543</th>
      <td>OSD-654</td>
      <td>2</td>
      <td>04-Oct-2023</td>
      <td>2023</td>
      <td>OsteoSys InSIGHT</td>
      <td>Dual-Energy X-Ray Absorptiometry</td>
      <td>Bone Microstructure</td>
      <td>1</td>
      <td>Mus musculus</td>
      <td>Spaceflight;Space Mission;Genotype</td>
      <td>non-omics</td>
    </tr>
    <tr>
      <th>544</th>
      <td>OSD-660</td>
      <td>1</td>
      <td>26-Sep-2023</td>
      <td>2023</td>
      <td>BioRad ChemiDoc</td>
      <td>Western Blot</td>
      <td>protein quantification</td>
      <td>1</td>
      <td>Mus musculus</td>
      <td>Spaceflight</td>
      <td>proteomics</td>
    </tr>
  </tbody>
</table>
<p>545 rows × 11 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Drop non-omics rows</span>
<span class="n">indices_to_drop</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;Category&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;non-omics&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">index</span>
<span class="n">omics_data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">indices_to_drop</span><span class="p">)</span> <span class="c1"># data</span>

<span class="n">omics_category_counts</span> <span class="o">=</span> <span class="n">omics_data</span><span class="p">[</span><span class="s1">&#39;Factors&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>

<span class="c1"># table of unique omics and their counts</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Table of Factors counts:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">omics_category_counts</span><span class="p">)</span>

<span class="c1">#  plot pie chart</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">pie</span><span class="p">(</span><span class="n">omics_category_counts</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">omics_category_counts</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">autopct</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%1.1f%%</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">startangle</span><span class="o">=</span><span class="mi">140</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Distribution of Omics&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Table of Factors counts:
Factors
Spaceflight                                        92
Spaceflight;Strain                                 13
Microgravity Simulation                            11
Spaceflight;Time                                   11
Spaceflight;Genotype                                8
                                                   ..
Treatment;Organism Part;Microgravity Simulation     1
Radiation;Timepoint                                 1
mouse strain;ionizing radiation                     1
ionizing radiation;Neoplasm                         1
Microgravity;Antibiotic concentration               1
Name: count, Length: 237, dtype: int64
</pre></div>
</div>
<img alt="_images/8de633b0663a319c1e6d8c12b74a3ae9f426105b25ff1559a1cd7f242ce11335.png" src="_images/8de633b0663a319c1e6d8c12b74a3ae9f426105b25ff1559a1cd7f242ce11335.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">omics_data</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>OSD-id</th>
      <th>hits</th>
      <th>publicReleaseDate</th>
      <th>publicReleaseDateYear</th>
      <th>technologyPlatform</th>
      <th>technologyType</th>
      <th>measurementType</th>
      <th>OrganismCount</th>
      <th>Organism</th>
      <th>Factors</th>
      <th>Category</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>OSD-1</td>
      <td>1</td>
      <td>11-Dec-2013</td>
      <td>2013</td>
      <td>Affymetrix</td>
      <td>microarray</td>
      <td>transcription profiling</td>
      <td>1</td>
      <td>Drosophila melanogaster</td>
      <td>Spaceflight;Infection</td>
      <td>transcriptomics</td>
    </tr>
    <tr>
      <th>1</th>
      <td>OSD-2</td>
      <td>1</td>
      <td>01-Apr-2011</td>
      <td>2011</td>
      <td>Agilent</td>
      <td>microarray</td>
      <td>transcription profiling</td>
      <td>1</td>
      <td>Homo sapiens</td>
      <td>Ionizing Radiation;Absorbed Radiation Dose</td>
      <td>transcriptomics</td>
    </tr>
    <tr>
      <th>2</th>
      <td>OSD-3</td>
      <td>1</td>
      <td>01-Sep-2011</td>
      <td>2011</td>
      <td>Affymetrix</td>
      <td>microarray</td>
      <td>transcription profiling</td>
      <td>1</td>
      <td>Drosophila melanogaster</td>
      <td>Developmental Stage;Spaceflight;Cosmic Radiation</td>
      <td>transcriptomics</td>
    </tr>
    <tr>
      <th>3</th>
      <td>OSD-4</td>
      <td>1</td>
      <td>10-Mar-2010</td>
      <td>2010</td>
      <td>Affymetrix</td>
      <td>microarray</td>
      <td>transcription profiling</td>
      <td>1</td>
      <td>Mus musculus</td>
      <td>Spaceflight</td>
      <td>transcriptomics</td>
    </tr>
    <tr>
      <th>4</th>
      <td>OSD-5</td>
      <td>1</td>
      <td>09-Feb-2006</td>
      <td>2006</td>
      <td>Affymetrix</td>
      <td>microarray</td>
      <td>transcription profiling</td>
      <td>1</td>
      <td>Homo sapiens</td>
      <td>simulated microgravity</td>
      <td>transcriptomics</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>529</th>
      <td>OSD-612</td>
      <td>2</td>
      <td>01-Aug-2023</td>
      <td>2023</td>
      <td>Illumina</td>
      <td>single-cell RNA sequencing</td>
      <td>transcription profiling</td>
      <td>1</td>
      <td>Mus musculus</td>
      <td>Spaceflight</td>
      <td>transcriptomics</td>
    </tr>
    <tr>
      <th>530</th>
      <td>OSD-612</td>
      <td>2</td>
      <td>01-Aug-2023</td>
      <td>2023</td>
      <td>Illumina</td>
      <td>single-cell ATAC-seq</td>
      <td>Chromatin Accessibility</td>
      <td>1</td>
      <td>Mus musculus</td>
      <td>Spaceflight</td>
      <td>epigenomics</td>
    </tr>
    <tr>
      <th>531</th>
      <td>OSD-613</td>
      <td>2</td>
      <td>01-Aug-2023</td>
      <td>2023</td>
      <td>Illumina</td>
      <td>single-cell RNA sequencing</td>
      <td>transcription profiling</td>
      <td>1</td>
      <td>Mus musculus</td>
      <td>Spaceflight;Age</td>
      <td>transcriptomics</td>
    </tr>
    <tr>
      <th>532</th>
      <td>OSD-613</td>
      <td>2</td>
      <td>01-Aug-2023</td>
      <td>2023</td>
      <td>Illumina</td>
      <td>single-cell ATAC-seq</td>
      <td>Chromatin Accessibility</td>
      <td>1</td>
      <td>Mus musculus</td>
      <td>Spaceflight;Age</td>
      <td>epigenomics</td>
    </tr>
    <tr>
      <th>544</th>
      <td>OSD-660</td>
      <td>1</td>
      <td>26-Sep-2023</td>
      <td>2023</td>
      <td>BioRad ChemiDoc</td>
      <td>Western Blot</td>
      <td>protein quantification</td>
      <td>1</td>
      <td>Mus musculus</td>
      <td>Spaceflight</td>
      <td>proteomics</td>
    </tr>
  </tbody>
</table>
<p>503 rows × 11 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">networkx</span> <span class="k">as</span> <span class="nn">nx</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">Counter</span>
<span class="kn">import</span> <span class="nn">random</span>

<span class="c1"># Sample rows of factors</span>
<span class="c1">#rows = [&quot;a;b;c&quot;, &quot;b;f&quot;, &quot;b&quot;, &quot;a;g&quot;, &quot;c;d;e&quot;, &quot;f;g&quot;, &quot;b;c&quot;, &quot;a;b&quot;]</span>
<span class="n">rows</span><span class="o">=</span><span class="n">omics_data</span><span class="p">[</span><span class="s1">&#39;Factors&#39;</span><span class="p">]</span>


<span class="c1"># Define a dictionary to perform replacements</span>
<span class="n">factor_replacements</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;time&quot;</span><span class="p">:</span> <span class="s2">&quot;Time&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Tiime&quot;</span><span class="p">:</span> <span class="s2">&quot;Time&quot;</span><span class="p">,</span>
    <span class="s2">&quot;absorbed radiation dose&quot;</span><span class="p">:</span> <span class="s2">&quot;Absorbed Radiation Dose&quot;</span><span class="p">,</span>
    <span class="s2">&quot;age&quot;</span><span class="p">:</span> <span class="s2">&quot;Age&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Gravity, Altered&quot;</span><span class="p">:</span> <span class="s2">&quot;Gravity&quot;</span><span class="p">,</span>
    <span class="c1"># Add more replacements as needed</span>
    <span class="s1">&#39;Bystander Effect&#39;</span><span class="p">:</span><span class="s1">&#39;Bystander&#39;</span><span class="p">,</span>
    <span class="s1">&#39;hindlimb unloading&#39;</span><span class="p">:</span> <span class="s1">&#39;Hindlimb unloading&#39;</span><span class="p">,</span>
    <span class="s1">&#39;Hindlimb Unloading&#39;</span><span class="p">:</span> <span class="s1">&#39;Hindlimb unloading&#39;</span><span class="p">,</span>
    <span class="s1">&#39;Ionzing Radiation&#39;</span><span class="p">:</span> <span class="s1">&#39;Ionizing Radiation&#39;</span><span class="p">,</span>
    <span class="s1">&#39;Radiation, Ionzing&#39;</span><span class="p">:</span> <span class="s1">&#39;Ionizing Radiation&#39;</span><span class="p">,</span>
    <span class="s1">&#39;Irradiated&#39;</span><span class="p">:</span> <span class="s1">&#39;Irradiation&#39;</span><span class="p">,</span>
    <span class="s1">&#39;ionizing radiation&#39;</span><span class="p">:</span> <span class="s1">&#39;Ionizing Radiation&#39;</span><span class="p">,</span>
    <span class="s1">&#39;light cycle&#39;</span><span class="p">:</span> <span class="s1">&#39;Light Cycle&#39;</span><span class="p">,</span>
    <span class="s1">&#39;organism part&#39;</span><span class="p">:</span> <span class="s1">&#39;Organism part&#39;</span><span class="p">,</span>
    <span class="s1">&#39;Organism Part&#39;</span><span class="p">:</span> <span class="s1">&#39;Organism part&#39;</span><span class="p">,</span>
    <span class="s1">&#39;environmentalstress&#39;</span><span class="p">:</span> <span class="s1">&#39;environmental stress&#39;</span><span class="p">,</span>
    <span class="s1">&#39;weightlessness&#39;</span><span class="p">:</span> <span class="s1">&#39;Weightlessness&#39;</span><span class="p">,</span>
    <span class="s1">&#39;timepoint&#39;</span><span class="p">:</span> <span class="s1">&#39;Timepoint&#39;</span><span class="p">,</span>
    <span class="s1">&#39;treatment&#39;</span><span class="p">:</span> <span class="s1">&#39;Treatment&#39;</span><span class="p">,</span>
    <span class="s1">&#39;temperature&#39;</span><span class="p">:</span> <span class="s1">&#39;Temperature&#39;</span><span class="p">,</span>
    <span class="s1">&#39;dose&#39;</span><span class="p">:</span> <span class="s1">&#39;Dose&#39;</span><span class="p">,</span>
    <span class="s1">&#39;strain&#39;</span><span class="p">:</span> <span class="s1">&#39;Strain&#39;</span><span class="p">,</span>
    <span class="s1">&#39;generation&#39;</span><span class="p">:</span> <span class="s1">&#39;Generation&#39;</span><span class="p">,</span>
    <span class="s1">&#39;genotype&#39;</span><span class="p">:</span> <span class="s1">&#39;Genotype&#39;</span><span class="p">,</span>
    <span class="s1">&#39;gravity&#39;</span><span class="p">:</span> <span class="s1">&#39;Gravity&#39;</span><span class="p">,</span>
    <span class="s1">&#39;gravitation&#39;</span><span class="p">:</span> <span class="s1">&#39;Gravity&#39;</span><span class="p">,</span>
    <span class="s1">&#39;radiation&#39;</span><span class="p">:</span> <span class="s1">&#39;Radiation&#39;</span><span class="p">,</span>
    <span class="s1">&#39;radiation dose&#39;</span><span class="p">:</span> <span class="s1">&#39;Radiation Dose&#39;</span><span class="p">,</span>
    <span class="s1">&#39;space flight&#39;</span><span class="p">:</span> <span class="s1">&#39;Spaceflight&#39;</span><span class="p">,</span>
    <span class="s1">&#39;spaceflight&#39;</span><span class="p">:</span> <span class="s1">&#39;Spaceflight&#39;</span><span class="p">,</span>
    <span class="s1">&#39;sex&#39;</span><span class="p">:</span> <span class="s1">&#39;Sex&#39;</span><span class="p">,</span>
    <span class="s1">&#39;simulated microgravity&#39;</span><span class="p">:</span> <span class="s1">&#39;Simulated Microgravity&#39;</span><span class="p">,</span>
    <span class="s1">&#39;Time of Sample Collection After Treatment&#39;</span><span class="p">:</span> <span class="s1">&#39;Time of sample collection after treatment&#39;</span><span class="p">,</span>
    <span class="s1">&#39;Time of Sample Collection after Treatment&#39;</span><span class="p">:</span> <span class="s1">&#39;Time of sample collection after treatment&#39;</span><span class="p">,</span>
    <span class="s1">&#39;generation number&#39;</span><span class="p">:</span> <span class="s1">&#39;Generation&#39;</span><span class="p">,</span>
    <span class="s1">&#39;hypergravity&#39;</span><span class="p">:</span> <span class="s1">&#39;Hypergravity&#39;</span><span class="p">,</span>
    <span class="s1">&#39;individual&#39;</span><span class="p">:</span> <span class="s1">&#39;Individual&#39;</span><span class="p">,</span>
    <span class="s1">&#39;mouse strain&#39;</span><span class="p">:</span> <span class="s1">&#39;Strain&#39;</span><span class="p">,</span>
    <span class="s1">&#39;Simulated microgravity&#39;</span><span class="p">:</span> <span class="s1">&#39;Simulated Microgravity&#39;</span>
<span class="p">}</span>

<span class="c1"># Function to apply replacements to a factor</span>
<span class="k">def</span> <span class="nf">apply_replacements</span><span class="p">(</span><span class="n">factor</span><span class="p">):</span>
    <span class="k">return</span> <span class="n">factor_replacements</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">factor</span><span class="p">,</span> <span class="n">factor</span><span class="p">)</span>

<span class="c1"># Preprocess the factors and create a new list</span>
<span class="n">preprocessed_factors</span> <span class="o">=</span> <span class="p">[</span><span class="n">apply_replacements</span><span class="p">(</span><span class="n">factor</span><span class="p">)</span> <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">rows</span> <span class="k">for</span> <span class="n">factor</span> <span class="ow">in</span> <span class="n">row</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)]</span>

<span class="c1"># Count preprocessed factor occurrences</span>
<span class="n">factor_counts</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span><span class="n">preprocessed_factors</span><span class="p">)</span>

<span class="c1"># Initialize a directed graph</span>
<span class="n">knowledge_graph</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">DiGraph</span><span class="p">()</span>

<span class="c1"># Set a threshold for minimum occurrences</span>
<span class="n">min_occurrences</span> <span class="o">=</span> <span class="mi">20</span>

<span class="c1"># Split the rows and add nodes and edges to the graph for factors that meet the threshold</span>
<span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">rows</span><span class="p">:</span>
    <span class="n">factors</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">factors</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="c1"># Add edges for relationships if both factors meet the threshold</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">factors</span><span class="p">)):</span>
            <span class="n">factor1</span> <span class="o">=</span> <span class="n">apply_replacements</span><span class="p">(</span><span class="n">factors</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
            <span class="n">factor2</span> <span class="o">=</span> <span class="n">apply_replacements</span><span class="p">(</span><span class="n">factors</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
            <span class="k">if</span> <span class="n">factor_counts</span><span class="p">[</span><span class="n">factor1</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_occurrences</span> <span class="ow">and</span> <span class="n">factor_counts</span><span class="p">[</span><span class="n">factor2</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_occurrences</span><span class="p">:</span>
                <span class="n">knowledge_graph</span><span class="o">.</span><span class="n">add_edge</span><span class="p">(</span><span class="n">factor1</span><span class="p">,</span> <span class="n">factor2</span><span class="p">,</span> <span class="n">connection_count</span><span class="o">=</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>  <span class="c1"># Assign a random connection count (you can replace this with your actual data)</span>

    <span class="c1"># Add nodes for individual factors that meet the threshold</span>
    <span class="k">for</span> <span class="n">factor</span> <span class="ow">in</span> <span class="n">factors</span><span class="p">:</span>
        <span class="n">preprocessed_factor</span> <span class="o">=</span> <span class="n">apply_replacements</span><span class="p">(</span><span class="n">factor</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">factor_counts</span><span class="p">[</span><span class="n">preprocessed_factor</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_occurrences</span><span class="p">:</span>
            <span class="n">knowledge_graph</span><span class="o">.</span><span class="n">add_node</span><span class="p">(</span><span class="n">preprocessed_factor</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">factor_counts</span><span class="p">[</span><span class="n">preprocessed_factor</span><span class="p">]</span> <span class="o">*</span> <span class="mi">40</span><span class="p">)</span>  <span class="c1"># Adjust the scaling factor for node size</span>

<span class="c1"># Extract node sizes and edge connection counts</span>
<span class="n">node_sizes</span> <span class="o">=</span> <span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;size&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">data</span> <span class="ow">in</span> <span class="n">knowledge_graph</span><span class="o">.</span><span class="n">nodes</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">True</span><span class="p">)]</span>
<span class="n">edge_connection_counts</span> <span class="o">=</span> <span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;connection_count&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">data</span> <span class="ow">in</span> <span class="n">knowledge_graph</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">True</span><span class="p">)]</span>

<span class="c1"># Use kamada_kawai_layout to ensure non-overlapping nodes</span>
<span class="n">pos</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">kamada_kawai_layout</span><span class="p">(</span><span class="n">knowledge_graph</span><span class="p">)</span>

<span class="c1"># You can color the arrows based on the connection count</span>
<span class="n">edge_colors</span> <span class="o">=</span> <span class="p">[</span><span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">plasma</span><span class="p">(</span><span class="mf">0.2</span> <span class="o">+</span> <span class="mf">0.2</span> <span class="o">*</span> <span class="n">count</span><span class="p">)</span> <span class="k">for</span> <span class="n">count</span> <span class="ow">in</span> <span class="n">edge_connection_counts</span><span class="p">]</span>  <span class="c1"># Adjust the colormap and range as needed</span>

<span class="c1"># Create a color bar legend</span>
<span class="n">sm</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">ScalarMappable</span><span class="p">(</span><span class="n">cmap</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">plasma</span><span class="p">,</span> <span class="n">norm</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">Normalize</span><span class="p">(</span><span class="n">vmin</span><span class="o">=</span><span class="nb">min</span><span class="p">(</span><span class="n">edge_connection_counts</span><span class="p">),</span> <span class="n">vmax</span><span class="o">=</span><span class="nb">max</span><span class="p">(</span><span class="n">edge_connection_counts</span><span class="p">)))</span>
<span class="n">sm</span><span class="o">.</span><span class="n">_A</span> <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># Fake the scalar mappable</span>


<span class="c1"># Create a figure with specific dimensions and adjust the margins</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">subplots_adjust</span><span class="p">(</span><span class="n">left</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span> <span class="n">top</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span> <span class="n">bottom</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>  <span class="c1"># Adjust the margins as needed</span>

<span class="c1"># Visualize the graph with non-overlapping nodes, reduced text size, and colored arrows</span>
<span class="n">nx</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">knowledge_graph</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">with_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">node_size</span><span class="o">=</span><span class="n">node_sizes</span><span class="p">,</span> <span class="n">node_color</span><span class="o">=</span><span class="s1">&#39;lightblue&#39;</span><span class="p">,</span> <span class="n">font_size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">font_color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">font_weight</span><span class="o">=</span><span class="s1">&#39;bold&#39;</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">edge_color</span><span class="o">=</span><span class="n">edge_colors</span><span class="p">,</span> <span class="n">arrows</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">edge_cmap</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">plasma</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">sm</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Factor connection counts&#39;</span><span class="p">,</span> <span class="n">shrink</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>

<span class="c1"># Create a custom legend for node sizes</span>
<span class="n">sizes</span> <span class="o">=</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">12</span><span class="p">]</span>  <span class="c1"># Define the sizes you want to show in the legend</span>
<span class="n">legend_labels</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Size 1&quot;</span><span class="p">,</span> <span class="s2">&quot;Size 2&quot;</span><span class="p">,</span> <span class="s2">&quot;Size 3&quot;</span><span class="p">]</span>  <span class="c1"># Labels for each size</span>

<span class="c1">## Create a legend for node sizes</span>
<span class="c1">#legend_handles = [plt.Line2D([0], [0], marker=&#39;o&#39;, color=&#39;w&#39;, markerfacecolor=&#39;lightblue&#39;, markersize=size, label=label) for size, label in zip(sizes, legend_labels)]</span>
<span class="c1">#plt.legend(handles=legend_handles, title=&#39;Node Sizes&#39;, loc=&#39;upper right&#39;)</span>


<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;knowledge graph-Factors-Omics.pdf&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;pdf&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/var/folders/94/rzsrkvpd7xv9p8nh_3hblqv40000gp/T/ipykernel_8823/91923518.py:110: MatplotlibDeprecationWarning: Unable to determine Axes to steal space for Colorbar. Using gca(), but will raise in the future. Either provide the *cax* argument to use as the Axes for the Colorbar, provide the *ax* argument to steal space from it, or add *mappable* to an Axes.
  plt.colorbar(sm, label=&#39;Factor connection counts&#39;, shrink=0.5)
</pre></div>
</div>
<img alt="_images/0e55f5856f94e4425c808af0a822d6ca5ce3e5effd5ea1b3c9aecd3ea7fae53f.png" src="_images/0e55f5856f94e4425c808af0a822d6ca5ce3e5effd5ea1b3c9aecd3ea7fae53f.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#print unique factors</span>

<span class="n">uni</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">preprocessed_factors</span><span class="p">)</span>

<span class="k">for</span> <span class="n">element</span> <span class="ow">in</span> <span class="n">uni</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">element</span><span class="p">)</span>

<span class="c1">#end</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Absorbed Radiation Dose
Accession
Age
Altered Gravity
Altered Gravity Simulator
Altitude
Antibiotic concentration
Atmospheric Composition
Atmospheric Pressure
Bed Rest
Bleomycin Treatment
Bystander
CO2 Concentration
Carcass Preservation Method
Cell Line
Cell cycle phase
Clinical Treatment
Collection Date
Cosmic Radiation
Cultivar
Culture Condition
Developmental Condition
Developmental Stage
Diet
Dissection Condition
Dissection Timeline
Donor
Dose
Dry Immersion
Duration
Ecotype
Electromagnetic Fields
Euthanasia Location
Euthanasia Method
Exercise
Experimental Campaign
Fractionated Dose
Freezing
Gender
Generation
Genotype
Gravity
Ground Based Facility
Growth Cycle
Growth Environment
Growth Time
HZE
Hindlimb unloading
Hypergravity
Individual
Infection
Intracranial Pressure
Ionizing Radiation
Irradiation
Library Construction Protocol
Light
Light Cycle
Magnetic Field
Microgravity
Microgravity Simulation
Mitochondrial Status
Muscle, Skeletal
Neoplasm
Nutrition
O2 Level
Order of Preservation
Organism part
Parabola
Partial Weight Bearing
Particle Charge
Preservation Method
Rad9 Status
Radiation
Radiation Distance
Radiation Dose
Relative Humidity
Sample Collection Protocol
Sample Condition
Sample Location
Sample Preservation Method
Sample Storage
Sample Storage Time
Sampling Time
Seed Sanitization
Sex
Sieve Size
Simulated Microgravity
Smoking Status
Space Mission
Spacecraft
Spaceflight
Spike-in Quality Control
Strain
Study Subject
Temperature
Time
Time after treatment
Time of sample collection after treatment
Timepoint
Tissue
Tissue Homogenate Preservation Time at -80C in RLT Buffer
Tissue Model
Tissue Preservation Method
Tissue Type
Tissue segment
Treatment
Treatment Duration
UV Mutagenesis
Weightlessness
Weightlessness Simulation
animal housing
antibody
cage
collection set
culture media
environment exposure
environmental stress
exposure duration
food deprivation
location
post radiation timepoint
pressure
radiation type
soil environment exposure
stress
tetanus toxoid
treatment group
vaccine adjuvant
variation
water deprivation
zygosity
</pre></div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="PubMed_scraping_notenook_for_ISSOP_figures.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">&lt;no title&gt;</p>
      </div>
    </a>
    <a class="right-next"
       href="SUMMARY.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Table of contents</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="simple visible nav section-nav flex-column">
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Richard Barker and the OSDR team
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>