{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[{"file_id":"176UGdwX5vCbL2_lFsLokHF6_wzZCMr5I","timestamp":1709564183957},{"file_id":"1jb2odfVhZjH6_6hPrXLafCyHaEYkZyJu","timestamp":1707250131392},{"file_id":"1bUDcgWr6vSQLioLfWtMIA15xpdKYZuUN","timestamp":1706036000814}],"collapsed_sections":["gMEmAFI5sAuC"]},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["# AI/ML methods notebook\n","\n"],"metadata":{"id":"QG3X-tGIpkMA"}},{"cell_type":"markdown","source":["# install Python packages\n","This notebook is equipped with a dedicated login shell, tailored to the environment in which it is executed. If you are utilizing your personal compute system, such as a laptop, the login corresponds to your individual compute system login. Conversely, when running this notebook on Google Colab, the login is attributed to the root user. The initiation of Linux shell commands within Jupyter notebook code cells is denoted by a preceding exclamation point (!).\n","\n","In the code cell below, the provided pip commands are employed to install a range of Python libraries essential for the tasks covered in this notebook. It's worth noting that additional Python libraries are automatically installed within our virtual environment."],"metadata":{"id":"gMEmAFI5sAuC"}},{"cell_type":"markdown","source":[],"metadata":{"id":"XPbm2EcyIs_i"}},{"cell_type":"code","source":["!pip install scikit-learn --no-cache\n","!pip install scanpy --no-cache\n","!pip install gseapy --no-cache\n","!pip install pydeseq2 --no-cache\n","!pip install pybiomart --no-cache\n","!pip install mygene --no-cache\n","!pip install sklearn_som  --no-cache\n","!pip install pandas --no-cache\n","!pip install numpy --no-cache\n","!pip install matplotlib --no-cache\n","!pip install sklearn-som --no-cache\n","!pip install pyDeseq2 --no-cache"],"metadata":{"id":"bb8-MziuOeFs","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1712588111860,"user_tz":300,"elapsed":104746,"user":{"displayName":"Richard Barker","userId":"08066349230655107502"}},"outputId":"f2a20662-25de-43a9-c102-af9a808f238c"},"execution_count":1,"outputs":[{"output_type":"stream","name":"stdout","text":["Requirement already satisfied: scikit-learn in /usr/local/lib/python3.10/dist-packages (1.2.2)\n","Requirement already satisfied: numpy>=1.17.3 in /usr/local/lib/python3.10/dist-packages (from scikit-learn) (1.25.2)\n","Requirement already satisfied: scipy>=1.3.2 in /usr/local/lib/python3.10/dist-packages (from scikit-learn) (1.11.4)\n","Requirement already satisfied: joblib>=1.1.1 in /usr/local/lib/python3.10/dist-packages (from scikit-learn) (1.3.2)\n","Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.10/dist-packages (from scikit-learn) (3.4.0)\n","Collecting scanpy\n","  Downloading scanpy-1.10.0-py3-none-any.whl (2.1 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.1/2.1 MB\u001b[0m \u001b[31m10.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting anndata>=0.8 (from scanpy)\n","  Downloading anndata-0.10.6-py3-none-any.whl (122 kB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m122.1/122.1 kB\u001b[0m \u001b[31m153.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: h5py>=3.1 in /usr/local/lib/python3.10/dist-packages (from scanpy) (3.9.0)\n","Requirement already satisfied: joblib in /usr/local/lib/python3.10/dist-packages (from scanpy) (1.3.2)\n","Collecting legacy-api-wrap>=1.4 (from scanpy)\n","  Downloading legacy_api_wrap-1.4-py3-none-any.whl (15 kB)\n","Requirement already satisfied: matplotlib>=3.6 in /usr/local/lib/python3.10/dist-packages (from scanpy) (3.7.1)\n","Requirement already satisfied: natsort in /usr/local/lib/python3.10/dist-packages (from scanpy) (8.4.0)\n","Requirement already satisfied: networkx>=2.7 in /usr/local/lib/python3.10/dist-packages (from scanpy) (3.2.1)\n","Requirement already satisfied: numba>=0.56 in /usr/local/lib/python3.10/dist-packages (from scanpy) (0.58.1)\n","Requirement already satisfied: numpy>=1.23 in /usr/local/lib/python3.10/dist-packages (from scanpy) (1.25.2)\n","Requirement already satisfied: packaging>=21.3 in /usr/local/lib/python3.10/dist-packages (from scanpy) (24.0)\n","Requirement already satisfied: pandas>=1.5 in /usr/local/lib/python3.10/dist-packages (from scanpy) (2.0.3)\n","Requirement already satisfied: patsy in /usr/local/lib/python3.10/dist-packages (from scanpy) (0.5.6)\n","Collecting pynndescent>=0.5 (from scanpy)\n","  Downloading pynndescent-0.5.12-py3-none-any.whl (56 kB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m56.8/56.8 kB\u001b[0m \u001b[31m127.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: scikit-learn>=0.24 in /usr/local/lib/python3.10/dist-packages (from scanpy) (1.2.2)\n","Requirement already satisfied: scipy>=1.8 in /usr/local/lib/python3.10/dist-packages (from scanpy) (1.11.4)\n","Requirement already satisfied: seaborn>=0.13 in /usr/local/lib/python3.10/dist-packages (from scanpy) (0.13.1)\n","Collecting session-info (from scanpy)\n","  Downloading session_info-1.0.0.tar.gz (24 kB)\n","  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n","Requirement already satisfied: statsmodels>=0.13 in /usr/local/lib/python3.10/dist-packages (from scanpy) (0.14.1)\n","Requirement already satisfied: tqdm in /usr/local/lib/python3.10/dist-packages (from scanpy) (4.66.2)\n","Collecting umap-learn!=0.5.0,>=0.5 (from scanpy)\n","  Downloading umap_learn-0.5.6-py3-none-any.whl (85 kB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m85.7/85.7 kB\u001b[0m \u001b[31m113.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting array-api-compat!=1.5,>1.4 (from anndata>=0.8->scanpy)\n","  Downloading array_api_compat-1.6-py3-none-any.whl (36 kB)\n","Requirement already satisfied: exceptiongroup in /usr/local/lib/python3.10/dist-packages (from anndata>=0.8->scanpy) (1.2.0)\n","Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6->scanpy) (1.2.0)\n","Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6->scanpy) (0.12.1)\n","Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6->scanpy) (4.50.0)\n","Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6->scanpy) (1.4.5)\n","Requirement already satisfied: pillow>=6.2.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6->scanpy) (9.4.0)\n","Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6->scanpy) (3.1.2)\n","Requirement already satisfied: python-dateutil>=2.7 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6->scanpy) (2.8.2)\n","Requirement already satisfied: llvmlite<0.42,>=0.41.0dev0 in /usr/local/lib/python3.10/dist-packages (from numba>=0.56->scanpy) (0.41.1)\n","Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas>=1.5->scanpy) (2023.4)\n","Requirement already satisfied: tzdata>=2022.1 in /usr/local/lib/python3.10/dist-packages (from pandas>=1.5->scanpy) (2024.1)\n","Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.10/dist-packages (from scikit-learn>=0.24->scanpy) (3.4.0)\n","Requirement already satisfied: six in /usr/local/lib/python3.10/dist-packages (from patsy->scanpy) (1.16.0)\n","Collecting stdlib_list (from session-info->scanpy)\n","  Downloading stdlib_list-0.10.0-py3-none-any.whl (79 kB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m79.8/79.8 kB\u001b[0m \u001b[31m135.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hBuilding wheels for collected packages: session-info\n","  Building wheel for session-info (setup.py) ... \u001b[?25l\u001b[?25hdone\n","  Created wheel for session-info: filename=session_info-1.0.0-py3-none-any.whl size=8026 sha256=aae3db65ed9cede1a3d6c1348e0e0bcd3956e56a01a20dcb0aeab59c293d187f\n","  Stored in directory: /tmp/pip-ephem-wheel-cache-vf_rt35b/wheels/6a/aa/b9/eb5d4031476ec10802795b97ccf937b9bd998d68a9b268765a\n","Successfully built session-info\n","Installing collected packages: array-api-compat, stdlib_list, legacy-api-wrap, session-info, pynndescent, anndata, umap-learn, scanpy\n","Successfully installed anndata-0.10.6 array-api-compat-1.6 legacy-api-wrap-1.4 pynndescent-0.5.12 scanpy-1.10.0 session-info-1.0.0 stdlib_list-0.10.0 umap-learn-0.5.6\n","Collecting gseapy\n","  Downloading gseapy-1.1.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.3 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.3/1.3 MB\u001b[0m \u001b[31m10.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: numpy>=1.13.0 in /usr/local/lib/python3.10/dist-packages (from gseapy) (1.25.2)\n","Requirement already satisfied: scipy in /usr/local/lib/python3.10/dist-packages (from gseapy) (1.11.4)\n","Requirement already satisfied: pandas in /usr/local/lib/python3.10/dist-packages (from gseapy) (2.0.3)\n","Requirement already satisfied: matplotlib>=2.2 in /usr/local/lib/python3.10/dist-packages (from gseapy) (3.7.1)\n","Requirement already satisfied: requests in /usr/local/lib/python3.10/dist-packages (from gseapy) (2.31.0)\n","Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=2.2->gseapy) (1.2.0)\n","Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=2.2->gseapy) (0.12.1)\n","Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=2.2->gseapy) (4.50.0)\n","Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=2.2->gseapy) (1.4.5)\n","Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=2.2->gseapy) (24.0)\n","Requirement already satisfied: pillow>=6.2.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=2.2->gseapy) (9.4.0)\n","Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=2.2->gseapy) (3.1.2)\n","Requirement already satisfied: python-dateutil>=2.7 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=2.2->gseapy) (2.8.2)\n","Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas->gseapy) (2023.4)\n","Requirement already satisfied: tzdata>=2022.1 in /usr/local/lib/python3.10/dist-packages (from pandas->gseapy) (2024.1)\n","Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests->gseapy) (3.3.2)\n","Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests->gseapy) (3.6)\n","Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests->gseapy) (2.0.7)\n","Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests->gseapy) (2024.2.2)\n","Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.7->matplotlib>=2.2->gseapy) (1.16.0)\n","Installing collected packages: gseapy\n","Successfully installed gseapy-1.1.2\n","Collecting pydeseq2\n","  Downloading pydeseq2-0.4.8-py3-none-any.whl (44 kB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m44.2/44.2 kB\u001b[0m \u001b[31m2.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: anndata>=0.8.0 in /usr/local/lib/python3.10/dist-packages (from pydeseq2) (0.10.6)\n","Requirement already satisfied: numpy>=1.23.0 in /usr/local/lib/python3.10/dist-packages (from pydeseq2) (1.25.2)\n","Requirement already satisfied: pandas>=1.4.0 in /usr/local/lib/python3.10/dist-packages (from pydeseq2) (2.0.3)\n","Requirement already satisfied: scikit-learn>=1.1.0 in /usr/local/lib/python3.10/dist-packages (from pydeseq2) (1.2.2)\n","Requirement already satisfied: scipy>=1.11.0 in /usr/local/lib/python3.10/dist-packages (from pydeseq2) (1.11.4)\n","Requirement already satisfied: matplotlib>=3.6.2 in /usr/local/lib/python3.10/dist-packages (from pydeseq2) (3.7.1)\n","Requirement already satisfied: array-api-compat!=1.5,>1.4 in /usr/local/lib/python3.10/dist-packages (from anndata>=0.8.0->pydeseq2) (1.6)\n","Requirement already satisfied: exceptiongroup in /usr/local/lib/python3.10/dist-packages (from anndata>=0.8.0->pydeseq2) (1.2.0)\n","Requirement already satisfied: h5py>=3.1 in /usr/local/lib/python3.10/dist-packages (from anndata>=0.8.0->pydeseq2) (3.9.0)\n","Requirement already satisfied: natsort in /usr/local/lib/python3.10/dist-packages (from anndata>=0.8.0->pydeseq2) (8.4.0)\n","Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/dist-packages (from anndata>=0.8.0->pydeseq2) (24.0)\n","Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6.2->pydeseq2) (1.2.0)\n","Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6.2->pydeseq2) (0.12.1)\n","Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6.2->pydeseq2) (4.50.0)\n","Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6.2->pydeseq2) (1.4.5)\n","Requirement already satisfied: pillow>=6.2.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6.2->pydeseq2) (9.4.0)\n","Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6.2->pydeseq2) (3.1.2)\n","Requirement already satisfied: python-dateutil>=2.7 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6.2->pydeseq2) (2.8.2)\n","Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas>=1.4.0->pydeseq2) (2023.4)\n","Requirement already satisfied: tzdata>=2022.1 in /usr/local/lib/python3.10/dist-packages (from pandas>=1.4.0->pydeseq2) (2024.1)\n","Requirement already satisfied: joblib>=1.1.1 in /usr/local/lib/python3.10/dist-packages (from scikit-learn>=1.1.0->pydeseq2) (1.3.2)\n","Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.10/dist-packages (from scikit-learn>=1.1.0->pydeseq2) (3.4.0)\n","Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.7->matplotlib>=3.6.2->pydeseq2) (1.16.0)\n","Installing collected packages: pydeseq2\n","Successfully installed pydeseq2-0.4.8\n","Collecting pybiomart\n","  Downloading pybiomart-0.2.0-py3-none-any.whl (10 kB)\n","Requirement already satisfied: future in /usr/local/lib/python3.10/dist-packages (from pybiomart) (0.18.3)\n","Requirement already satisfied: pandas in /usr/local/lib/python3.10/dist-packages (from pybiomart) (2.0.3)\n","Requirement already satisfied: requests in /usr/local/lib/python3.10/dist-packages (from pybiomart) (2.31.0)\n","Collecting requests-cache (from pybiomart)\n","  Downloading requests_cache-1.2.0-py3-none-any.whl (61 kB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m61.4/61.4 kB\u001b[0m \u001b[31m3.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.10/dist-packages (from pandas->pybiomart) (2.8.2)\n","Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas->pybiomart) (2023.4)\n","Requirement already satisfied: tzdata>=2022.1 in /usr/local/lib/python3.10/dist-packages (from pandas->pybiomart) (2024.1)\n","Requirement already satisfied: numpy>=1.21.0 in /usr/local/lib/python3.10/dist-packages (from pandas->pybiomart) (1.25.2)\n","Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests->pybiomart) (3.3.2)\n","Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests->pybiomart) (3.6)\n","Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests->pybiomart) (2.0.7)\n","Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests->pybiomart) (2024.2.2)\n","Requirement already satisfied: attrs>=21.2 in /usr/local/lib/python3.10/dist-packages (from requests-cache->pybiomart) (23.2.0)\n","Collecting cattrs>=22.2 (from requests-cache->pybiomart)\n","  Downloading cattrs-23.2.3-py3-none-any.whl (57 kB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m57.5/57.5 kB\u001b[0m \u001b[31m24.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: platformdirs>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests-cache->pybiomart) (4.2.0)\n","Collecting url-normalize>=1.4 (from requests-cache->pybiomart)\n","  Downloading url_normalize-1.4.3-py2.py3-none-any.whl (6.8 kB)\n","Requirement already satisfied: exceptiongroup>=1.1.1 in /usr/local/lib/python3.10/dist-packages (from cattrs>=22.2->requests-cache->pybiomart) (1.2.0)\n","Requirement already satisfied: typing-extensions!=4.6.3,>=4.1.0 in /usr/local/lib/python3.10/dist-packages (from cattrs>=22.2->requests-cache->pybiomart) (4.10.0)\n","Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.8.2->pandas->pybiomart) (1.16.0)\n","Installing collected packages: url-normalize, cattrs, requests-cache, pybiomart\n","Successfully installed cattrs-23.2.3 pybiomart-0.2.0 requests-cache-1.2.0 url-normalize-1.4.3\n","Collecting mygene\n","  Downloading mygene-3.2.2-py2.py3-none-any.whl (5.4 kB)\n","Collecting biothings-client>=0.2.6 (from mygene)\n","  Downloading biothings_client-0.3.1-py2.py3-none-any.whl (29 kB)\n","Requirement already satisfied: requests>=2.3.0 in /usr/local/lib/python3.10/dist-packages (from biothings-client>=0.2.6->mygene) (2.31.0)\n","Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests>=2.3.0->biothings-client>=0.2.6->mygene) (3.3.2)\n","Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests>=2.3.0->biothings-client>=0.2.6->mygene) (3.6)\n","Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests>=2.3.0->biothings-client>=0.2.6->mygene) (2.0.7)\n","Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests>=2.3.0->biothings-client>=0.2.6->mygene) (2024.2.2)\n","Installing collected packages: biothings-client, mygene\n","Successfully installed biothings-client-0.3.1 mygene-3.2.2\n","Collecting sklearn_som\n","  Downloading sklearn_som-1.1.0-py3-none-any.whl (6.7 kB)\n","Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from sklearn_som) (1.25.2)\n","Installing collected packages: sklearn_som\n","Successfully installed sklearn_som-1.1.0\n","Requirement already satisfied: pandas in /usr/local/lib/python3.10/dist-packages (2.0.3)\n","Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.10/dist-packages (from pandas) (2.8.2)\n","Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas) (2023.4)\n","Requirement already satisfied: tzdata>=2022.1 in /usr/local/lib/python3.10/dist-packages (from pandas) (2024.1)\n","Requirement already satisfied: numpy>=1.21.0 in /usr/local/lib/python3.10/dist-packages (from pandas) (1.25.2)\n","Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.8.2->pandas) (1.16.0)\n","Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (1.25.2)\n","Requirement already satisfied: matplotlib in /usr/local/lib/python3.10/dist-packages (3.7.1)\n","Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (1.2.0)\n","Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (0.12.1)\n","Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (4.50.0)\n","Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (1.4.5)\n","Requirement already satisfied: numpy>=1.20 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (1.25.2)\n","Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (24.0)\n","Requirement already satisfied: pillow>=6.2.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (9.4.0)\n","Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (3.1.2)\n","Requirement already satisfied: python-dateutil>=2.7 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (2.8.2)\n","Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.7->matplotlib) (1.16.0)\n","Requirement already satisfied: sklearn-som in /usr/local/lib/python3.10/dist-packages (1.1.0)\n","Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from sklearn-som) (1.25.2)\n","Requirement already satisfied: pyDeseq2 in /usr/local/lib/python3.10/dist-packages (0.4.8)\n","Requirement already satisfied: anndata>=0.8.0 in /usr/local/lib/python3.10/dist-packages (from pyDeseq2) (0.10.6)\n","Requirement already satisfied: numpy>=1.23.0 in /usr/local/lib/python3.10/dist-packages (from pyDeseq2) (1.25.2)\n","Requirement already satisfied: pandas>=1.4.0 in /usr/local/lib/python3.10/dist-packages (from pyDeseq2) (2.0.3)\n","Requirement already satisfied: scikit-learn>=1.1.0 in /usr/local/lib/python3.10/dist-packages (from pyDeseq2) (1.2.2)\n","Requirement already satisfied: scipy>=1.11.0 in /usr/local/lib/python3.10/dist-packages (from pyDeseq2) (1.11.4)\n","Requirement already satisfied: matplotlib>=3.6.2 in /usr/local/lib/python3.10/dist-packages (from pyDeseq2) (3.7.1)\n","Requirement already satisfied: array-api-compat!=1.5,>1.4 in /usr/local/lib/python3.10/dist-packages (from anndata>=0.8.0->pyDeseq2) (1.6)\n","Requirement already satisfied: exceptiongroup in /usr/local/lib/python3.10/dist-packages (from anndata>=0.8.0->pyDeseq2) (1.2.0)\n","Requirement already satisfied: h5py>=3.1 in /usr/local/lib/python3.10/dist-packages (from anndata>=0.8.0->pyDeseq2) (3.9.0)\n","Requirement already satisfied: natsort in /usr/local/lib/python3.10/dist-packages (from anndata>=0.8.0->pyDeseq2) (8.4.0)\n","Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/dist-packages (from anndata>=0.8.0->pyDeseq2) (24.0)\n","Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6.2->pyDeseq2) (1.2.0)\n","Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6.2->pyDeseq2) (0.12.1)\n","Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6.2->pyDeseq2) (4.50.0)\n","Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6.2->pyDeseq2) (1.4.5)\n","Requirement already satisfied: pillow>=6.2.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6.2->pyDeseq2) (9.4.0)\n","Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6.2->pyDeseq2) (3.1.2)\n","Requirement already satisfied: python-dateutil>=2.7 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6.2->pyDeseq2) (2.8.2)\n","Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas>=1.4.0->pyDeseq2) (2023.4)\n","Requirement already satisfied: tzdata>=2022.1 in /usr/local/lib/python3.10/dist-packages (from pandas>=1.4.0->pyDeseq2) (2024.1)\n","Requirement already satisfied: joblib>=1.1.1 in /usr/local/lib/python3.10/dist-packages (from scikit-learn>=1.1.0->pyDeseq2) (1.3.2)\n","Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.10/dist-packages (from scikit-learn>=1.1.0->pyDeseq2) (3.4.0)\n","Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.7->matplotlib>=3.6.2->pyDeseq2) (1.16.0)\n"]}]},{"cell_type":"markdown","source":["# import Python modules"],"metadata":{"id":"cpkDoGrUsMlL"}},{"cell_type":"markdown","source":["This notebook imports a number of Python modules for use in several notebooks."],"metadata":{"id":"uSnhzxTaLN9l"}},{"cell_type":"code","source":["import requests\n","import json\n","import pandas as pd\n","from urllib.request import urlretrieve\n","import numpy as np\n","from sklearn.decomposition import PCA\n","from sklearn.cluster import KMeans\n","from sklearn_som.som import SOM\n","import matplotlib.pyplot as plt\n","from matplotlib.colors import ListedColormap\n","import scanpy as sc\n","import gseapy as gp\n","from gseapy.plot import gseaplot\n","from pydeseq2.dds import DeseqDataSet\n","from pydeseq2.ds import DeseqStats\n","from gseapy import Msigdb\n","from pybiomart import Server\n","import mygene\n","import seaborn as sns\n","from sklearn.decomposition import PCA, FastICA\n","from sklearn.cluster import DBSCAN\n","from sklearn.preprocessing import StandardScaler\n","from sklearn.linear_model import LinearRegression\n","from sklearn.linear_model import Ridge\n","from sklearn import linear_model\n","from sklearn.linear_model import TweedieRegressor\n","from sklearn.model_selection import train_test_split\n","from sklearn import metrics\n","from math import log\n","import statsmodels.api as sm\n","import pylab\n","import operator\n","from sklearn.mixture import GaussianMixture\n","from sklearn.ensemble import RandomForestClassifier\n","from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, recall_score, ConfusionMatrixDisplay\n","from sklearn.tree import export_graphviz\n","from IPython.display import Image\n","import graphviz\n","from sklearn.linear_model import LogisticRegression\n","from sklearn.linear_model import Perceptron\n","from sklearn.neural_network import MLPClassifier\n","from sklearn.inspection import permutation_importance\n","from itertools import islice\n","from sklearn.metrics import r2_score\n","from sklearn.metrics import mean_absolute_error\n","from sklearn.metrics import mean_squared_error"],"metadata":{"id":"f079g46t5jTU","executionInfo":{"status":"ok","timestamp":1712588125101,"user_tz":300,"elapsed":13245,"user":{"displayName":"Richard Barker","userId":"08066349230655107502"}}},"execution_count":2,"outputs":[]},{"cell_type":"markdown","source":["# define misc helper methods"],"metadata":{"id":"xxokeQF9rjFG"}},{"cell_type":"code","source":["def set_maxdisplay(n=None):\n","  pd.set_option('display.max_rows', n)\n","  from notebook.services.config import ConfigManager\n","  cm = ConfigManager().update('notebook', {'limit_output': n})"],"metadata":{"id":"EpOLwn3YqdYA","executionInfo":{"status":"ok","timestamp":1712588965698,"user_tz":300,"elapsed":400,"user":{"displayName":"Richard Barker","userId":"08066349230655107502"}}},"execution_count":3,"outputs":[]},{"cell_type":"markdown","source":["# Define data ingestion methods"],"metadata":{"id":"W_cXRde853Lu"}},{"cell_type":"markdown","source":[],"metadata":{"id":"SCSHct2Opf8X"}},{"cell_type":"code","source":["def read_meta_data(dataset):\n","  # dataset=255\n","  url = 'https://osdr.nasa.gov/geode-py/ws/studies/OSD-' + str(dataset) + '/download?source=datamanager&file=OSD-' + dataset + '_metadata_OSD-' + dataset + '-ISA.zip'\n","  filename = dataset + '-meta.zip'\n","  urlretrieve(url, filename)\n","  !unzip -o {filename} > /dev/null\n","  df = pd.read_csv('s_OSD-' + dataset + '.txt', sep='\\t', header=0)\n","  return df"],"metadata":{"id":"LwoDZ49r7i_6","executionInfo":{"status":"ok","timestamp":1712588967951,"user_tz":300,"elapsed":4,"user":{"displayName":"Richard Barker","userId":"08066349230655107502"}}},"execution_count":4,"outputs":[]},{"cell_type":"code","source":["def read_rnaseq_data(data):\n","  # data = '255_rna_seq_Normalized_Counts'\n","  dataset = data.split('_')[0]\n","  url='https://osdr.nasa.gov/geode-py/ws/studies/OSD-' + str(dataset) + '/download?source=datamanager&file=GLDS-' + data + '.csv'\n","  df = pd.read_csv(url)\n","  return df"],"metadata":{"id":"eOQxRBfUs6Ir","executionInfo":{"status":"ok","timestamp":1712588970232,"user_tz":300,"elapsed":254,"user":{"displayName":"Richard Barker","userId":"08066349230655107502"}}},"execution_count":5,"outputs":[]},{"cell_type":"code","source":["def read_phenotype_data(dataset, data):\n","  # dataset = '557'\n","  # data = 'LSDS-1_immunostaining_microscopy_PNAtr_Transformed_Reusable_Results'\n","  url='https://osdr.nasa.gov//geode-py/ws/studies/OSD-' + str(dataset) + '/download?source=datamanager&file=' + data + '.csv'\n","  df = pd.read_csv(url)\n","  return df"],"metadata":{"id":"yCVR-HkZn08K","executionInfo":{"status":"ok","timestamp":1712588971913,"user_tz":300,"elapsed":279,"user":{"displayName":"Richard Barker","userId":"08066349230655107502"}}},"execution_count":6,"outputs":[]},{"cell_type":"markdown","source":["# define data filtering methods"],"metadata":{"id":"7Ga9Pbi5rr1R"}},{"cell_type":"code","source":["def filter_cvs(df, frac=0.5):\n","\n","  # calculate coefficient of variation\n","  cvs=list()\n","  for i in range(len(df)):\n","    m=np.mean(df.iloc[i][1:])\n","    sd=np.std(df.iloc[i][1:])\n","    cvs.append(sd/m)\n","\n","  # plot hist of dist of coev of variation\n","  fig, axs = plt.subplots()\n","  axs.hist(cvs, bins=20)\n","\n","  # keep genes with cv > frac\n","  indices = list()\n","  for i in range(len(cvs)):\n","    if cvs[i] > frac:\n","      indices.append(i)\n","  return df.iloc[indices]\n"],"metadata":{"id":"vJaculrSRZp_","executionInfo":{"status":"ok","timestamp":1712588975356,"user_tz":300,"elapsed":309,"user":{"displayName":"Richard Barker","userId":"08066349230655107502"}}},"execution_count":7,"outputs":[]},{"cell_type":"code","source":["def drop_nans(df):\n","  # drop NaN rows\n","  df.dropna(inplace=True)\n","  return df\n"],"metadata":{"id":"0kMwy7X10hxR","executionInfo":{"status":"ok","timestamp":1712588977548,"user_tz":300,"elapsed":270,"user":{"displayName":"Richard Barker","userId":"08066349230655107502"}}},"execution_count":8,"outputs":[]},{"cell_type":"code","source":["def drop_lowcount(df, threshold=10):\n","\n","  # let's drop any low-count genes\n","  print(len(df))\n","  if 'transcript' in df.columns:\n","    df = df[df.drop(columns=['transcript']).sum(axis=1) >= threshold]\n","  elif 'Unnamed: 0' in df.columns:\n","    df = df[df.drop(columns=['Unnamed: 0']).sum(axis=1) >= threshold]\n","    df.rename(columns={\"Unnamed: 0\":\"transcript\"}, inplace=True)\n","  else:\n","    raise Exception(\"check file format\")\n","  return df\n"],"metadata":{"id":"t1tNC979OtBg","executionInfo":{"status":"ok","timestamp":1712588979547,"user_tz":300,"elapsed":267,"user":{"displayName":"Richard Barker","userId":"08066349230655107502"}}},"execution_count":9,"outputs":[]},{"cell_type":"code","source":["def filter_protein(df, drop='non-coding'):\n","  # let's filter protein/ non-protein-coding genes\n","  server = Server(host='http://www.ensembl.org')\n","  dataset = (server.marts['ENSEMBL_MART_ENSEMBL'].datasets['mmusculus_gene_ensembl'])\n","  gene_info = dataset.query(attributes=['ensembl_gene_id', 'external_gene_name', 'gene_biotype'])\n","  if drop=='non-coding':\n","    filter_genes=gene_info[gene_info['Gene type'] == 'protein_coding']['Gene stable ID']\n","  elif drop=='coding':\n","    filter_genes=gene_info[gene_info['Gene type'] != 'protein_coding']['Gene stable ID']\n","\n","  df=df[df['Unnamed: 0'].isin(filter_genes)]\n","  return df"],"metadata":{"id":"LegH26QSPhp7","executionInfo":{"status":"ok","timestamp":1712588980715,"user_tz":300,"elapsed":2,"user":{"displayName":"Richard Barker","userId":"08066349230655107502"}}},"execution_count":10,"outputs":[]},{"cell_type":"code","source":["def filter_data(df, dropnans=False, dropprotein='non-coding', droplowvar=0):\n","  # drop NANs\n","  if dropnans:\n","    df = drop_nans(df)\n","  # drop non protein-coding genes\n","  df = filter_protein(df, drop=dropprotein)\n","  # drop low variation genes\n","  if droplowvar != 0:\n","    df = filter_cvs(df, droplowvar)\n","  return df"],"metadata":{"id":"BIopMUVv2jpH","executionInfo":{"status":"ok","timestamp":1712588983285,"user_tz":300,"elapsed":734,"user":{"displayName":"Richard Barker","userId":"08066349230655107502"}}},"execution_count":11,"outputs":[]},{"cell_type":"code","source":["def exclude_samples_by_prefix(df, prefix=\"V\", colname=\"Source Name\"):\n","  sample_names=list(df[colname].values)\n","  exclude_names=list()\n","  for sn in sample_names:\n","    if sn.startswith(prefix):\n","      exclude_names.append(sn)\n","  return exclude_names"],"metadata":{"id":"7grY5F-aEG_6","executionInfo":{"status":"ok","timestamp":1712588984217,"user_tz":300,"elapsed":1,"user":{"displayName":"Richard Barker","userId":"08066349230655107502"}}},"execution_count":12,"outputs":[]},{"cell_type":"markdown","source":["# data transformation methods"],"metadata":{"id":"d4IuHC13rGZv"}},{"cell_type":"code","source":["def transpose_df(df, cur_index_col, new_index_col):\n","  df = df.set_index(cur_index_col).T\n","  df.reset_index(level=0, inplace=True)\n","  cols = [new_index_col] + list(df.columns)[1:]\n","  df.columns = cols\n","  return df"],"metadata":{"id":"JFOBNzQ-rIW4","executionInfo":{"status":"ok","timestamp":1712588986026,"user_tz":300,"elapsed":296,"user":{"displayName":"Richard Barker","userId":"08066349230655107502"}}},"execution_count":13,"outputs":[]},{"cell_type":"code","source":["def reduce_dims(df, current_key, new_key, n):\n","  #df_t = transpose_df(df, current_key, new_key)\n","  #sdList = df_t.var(axis=1)\n","  sdList = df.std(axis=1)\n","  print('len of sdlist: ', str(len(sdList)))\n","  sdDict = {k: v for v, k in enumerate(sdList)}\n","  if n < 0:\n","    sdDictSorted = sorted(sdDict.items(), key=operator.itemgetter(0), reverse=False)\n","  else:\n","    sdDictSorted = sorted(sdDict.items(), key=operator.itemgetter(0), reverse=True)\n","  topN = sdDictSorted[0:abs(n)]\n","  print('n: ', n)\n","  indices = [x[1] for x in topN]\n","  #df_t = df_t.iloc[indices]\n","  #df_tt= transpose_df(df_t, new_key, current_key)\n","  return df.iloc[indices]"],"metadata":{"id":"6jVaYnW9rRTU","executionInfo":{"status":"ok","timestamp":1712588986798,"user_tz":300,"elapsed":2,"user":{"displayName":"Richard Barker","userId":"08066349230655107502"}}},"execution_count":14,"outputs":[]},{"cell_type":"code","source":["def convert_pd_to_np(df):\n","  X=list()\n","  for col in df.columns[1:]:\n","    X.append(list(df[col]))\n","  return np.array(X)"],"metadata":{"id":"Kip2oJPk9ArZ","executionInfo":{"status":"ok","timestamp":1712588988665,"user_tz":300,"elapsed":2,"user":{"displayName":"Richard Barker","userId":"08066349230655107502"}}},"execution_count":15,"outputs":[]},{"cell_type":"markdown","source":["# plotting methods"],"metadata":{"id":"EYvmAZ627HKz"}},{"cell_type":"code","source":["import pandas as pd\n","import matplotlib.pyplot as plt\n","import argparse\n","import scipy.stats as stats\n","\n","def plotbox_and_stats(data_, sample_key, field, treatment, space, exclude_samples=[]):\n","  print('field: ', field)\n","  print('excluding samples: ', exclude_samples)\n","  fieldValues = set(data_[field])\n","  value_dict=dict()\n","  results = dict()\n","\n","  flight = str(field) + '_flight'\n","  nonflight= str(field) + '_nonflight'\n","  results[field] = dict()\n","  value_dict[flight] = list()\n","  value_dict[nonflight] = list()\n","  for i in range(len(data_)):\n","    if data_.iloc[i][sample_key] in exclude_samples:\n","      continue\n","    elif treatment is None:\n","      if data_.iloc[i][sample_key].startswith('F'):\n","        value_dict[flight].append(data_.iloc[i][field])\n","      else:\n","        value_dict[nonflight].append(data_.iloc[i][field])\n","    else:\n","      if data_.iloc[i][treatment] == space:\n","        value_dict[flight].append(data_.iloc[i][field])\n","      else:\n","        value_dict[nonflight].append(data_.iloc[i][field])\n","\n","\n","  if len(value_dict[flight]) != 0 and len(value_dict[nonflight]) != 0:\n","    results[field]['t-test p-value'] = float('%.5f' % (stats.ttest_ind(value_dict[flight], value_dict[nonflight]).pvalue))\n","    results[field]['wilcoxon p-value'] = float('%.5f' % (stats.ranksums(value_dict[flight], value_dict[nonflight]).pvalue))\n","    results[field]['ks-test p-value'] = float('%.5f' % (stats.kstest(value_dict[flight], value_dict[nonflight]).pvalue))\n","\n","\n","  print(results)\n","  print('n flight = ', len(value_dict[flight]))\n","  print('n nonflight = ', len(value_dict[nonflight]))\n","  fig,ax = plt.subplots()\n","  ax.boxplot(value_dict.values())\n","  ax.set_xticklabels(value_dict.keys())\n","  plt.show()"],"metadata":{"id":"EozbsWlf7Lf4","executionInfo":{"status":"ok","timestamp":1712588990574,"user_tz":300,"elapsed":227,"user":{"displayName":"Richard Barker","userId":"08066349230655107502"}}},"execution_count":16,"outputs":[]}]}